zfs/zfs.go

```diff
@@ -2,79 +2,85 @@ package zfs
 
 import (
 	"bufio"
+	"bytes"
 	"fmt"
 	"io"
+	"log"
 	"os/exec"
 	"strings"
 )
 
+// Zfs is a wrapper for local or remote ZFS commands
 type Zfs struct {
 	exec Exec
 }
 
-// NewLocal creates a wrapper for local ZFS commands
-func NewLocal() *Zfs {
-	return &Zfs{
-		exec: LocalExec,
+func ParseLocation(location string) (z *Zfs, fspath string) {
+	colon := strings.LastIndexByte(location, ':')
+
+	if colon == -1 {
+		z = &Zfs{
+			exec: LocalExec,
+		}
+		fspath = location
+	} else {
+		z = &Zfs{
+			exec: RemoteExecutor(location[:colon]),
+		}
+		fspath = location[colon+1:]
 	}
+
+	return
 }
 
-// NewRemote creates a wrapper for remote ZFS commands
-func NewRemote(host string, user string) *Zfs {
-	return &Zfs{
-		exec: RemoteExecutor(fmt.Sprintf("%s@%s", user, host)),
+func GetFilesystem(location string) (*Fs, error) {
+	z, fspath := ParseLocation(location)
+	fs, err := z.List()
+	if err != nil {
+		return nil, err
 	}
+
+	return fs.GetChild(fspath)
 }
 
 // List returns all ZFS volumes and snapshots
-func (z *Zfs) List() (Fs, error) {
-	cmd := z.exec("/sbin/zfs", "list", "-t", "all", "-o", "name")
+func (z *Zfs) List() (*Fs, error) {
+	cmd := z.exec("/sbin/zfs", "list", "-t", "all", "-Hr", "-o", "name")
 	b, err := cmd.Output()
 	if err != nil {
 		return nil, err
 	}
-	return z.parseList(b)
+	return z.parseList(b), nil
 }
 
-func (z *Zfs) parseList(b []byte) (Fs, error) {
-	s := string(b)
-	scanner := bufio.NewScanner(strings.NewReader(s))
-	var f Fs
-	scanner.Scan()
-
-	l := scanner.Text()
-	if l != "NAME" {
-		return nil, fmt.Errorf("First line should be NAME: %s", l)
-	}
+func (z *Zfs) parseList(b []byte) *Fs {
+	root := newFs(z, "")
+	scanner := bufio.NewScanner(bytes.NewReader(b))
 
 	for scanner.Scan() {
-		l := scanner.Text()
-		isSnap := strings.Contains(l, "@")
-		if f == nil {
-			if isSnap {
-				return nil, fmt.Errorf("First element should not be snapshot: %s", l)
-			}
-
-			f = NewFs(z, l)
+		line := scanner.Text()
+		if strings.ContainsRune(line, '@') {
+			root.addSnapshot(line)
 		} else {
-			if isSnap {
-				f.AddSnapshot(l)
-			} else {
-				f.AddChild(l)
-			}
+			root.addChild(line)
 		}
-
 	}
+	return root
+}
 
-	return f, nil
+// Create creates a new filesystem by its full path
+func (z *Zfs) Create(fs string) error {
+	_, err := z.exec("/sbin/zfs", "create", fs).Output()
+	return err
 }
 
+// Recv performs the `zfs recv` command
 func (z *Zfs) Recv(fs string, sendCommand *exec.Cmd) error {
-	cmd := z.exec("/sbin/zfs", "recv", fs)
+	cmd := z.exec("/sbin/zfs", "recv", "-F", fs)
```

Comment:
I see a "minor" problem here

The -F flag make it so that you can by mistake completely override a local filesystem with no snapshot by the remote one:

```bash
zfs create rpool/test
zfs create rpool/test/orig
zfs create zpool test/dest

# Create a different file in each of the filesystems
touch /test/orig/file.orig
touch /test/dest/file.dest

# Synchronize; will overwrite /test/dest with /test/orig
zfs-snapback sync rpool/test/orig rpool/test/dest

ls /test/dest
```

In itself, the behavior is not bad. What is not good is that it does not warn the user about the possibility to have an unwanted data loss.

Reply:
Maybe the solution would be to refuse to overwrite a local filesystem with no snapshot in common with the source if there is not a `-F` flag in the command line?

Reply:
Updated example.